Adding support for edge weights (using SimpleWeightedGraphs)

rocarvaj opened the pull request:
Hello! First of all, please note that I'm quite new to Julia and this is my first time contributing to a project.

Inspired by #35, I decided to modify the `partition` function in order to be able to use weights on edges. At first I thought of just adding an extra argument with the `adjwgt` vector with the weights, but this vector would have to be constructed by the user in the Metis CSR format which is not very friendly.
So, I decided to add a method  for  the `graph()` function that converts a weighted graph from [SimpleWeightedGraphs.jl](https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl) into a CSR graph.

A few things that I'm not sure about (and that I would be thankful to know your opinions on):

- I don't know how to add `SimpleWeightedGraphs` to the package's dependencies. Every time I do `using Metis` in the REPL I get the message: "Warning: Package Metis does not have SimpleWeightedGraphs in its dependencies".
- I couldn't find a way of having an inner constructor for `Graph` that would ignore the `vwgt` field but not `adjwgt`. What I ended up doing was setting `vwgt` to a vector of zeros (this should not affect the behavior of METIS).
- I don't know how to write tests for these changes.

Thanks in advance for the chance of contributing!

tsano430 commented:
This PR closes #35.

tsano430 commented:
@fredrikekre, could you please review this PR? I'd like to use the support for edge weights.

fredrikekre commented:
Superficially LGTM but I don't know anything about the graph library so would be good to get a review from someone active there (@jpfairbanks or @matbesancon maybe? sorry if you are the wrong people to ping here.)

 - SimpleWeightedGraphs needs to be added to dependencies in `Project.toml`
 - Please include some test.

rocarvaj commented:
I've added SimpleWeightedGraphs to the package dependencies. I noticed that setting the vertex weights to 0 results in weird behavior in METIS (I actually tested this running the METIS library directly in a separate installation) so now the `graph()` method for graphs with edge weights sets the vertex weights to 1.

I've added a simple test for the edge weight support. Following @fredrikekre 's suggestion of comparing the partition obtained from setting the edge weights to 1 with the partition obtained with the unweighted version.

matbesancon commented:
LGTM even though I don't have much knowledge of Metis. For the weight of 0, SimpleWG is a sparse matrix under the hood, so adding an edge with a weight of 0 will not add the edge I believe (there are checks to `dropzeros`).

rocarvaj commented:
> LGTM even though I don't have much knowledge of Metis. For the weight of 0, SimpleWG is a sparse matrix under the hood, so adding an edge with a weight of 0 will not add the edge I believe (there are checks to `dropzeros`).

I get what you say regarding edge weights, but I’m referring to *vertex weights*. I couldn’t construct a `Graph` struct initializing `adjwgt` (Edge weights) but not `vwgt` (Vertex weights), that’s why I ended up initializing the vertex weights to 1.

tsano430 commented on src/Metis.jl:
```diff
@@ -82,6 +85,40 @@ function graph(G::LightGraphs.AbstractSimpleGraph)
+    N = SimpleWeightedGraphs.nv(G)
+    xadj = Vector{idx_t}(undef, N+1)
+    xadj[1] = 1
+    adjncy = Vector{idx_t}(undef, 2*SimpleWeightedGraphs.ne(G))
```
There is a discrepancy between `SimpleWeightedGraphs.ne()` and `LightGraphs.ne()` because `SimpleWeightedGraphs.ne()` fails to count the correct number of self-loops (please see [here](https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl/issues/41)). The current `Metis.partition()` is able to handle unweighted graphs with self-loops, so I think this part should be fixed to allow self-loops.

tsano430 replied:
At least the following implementation avoids the problem.

```julia
adjncy = Vector{idx_t}(undef, 2*(SimpleWeightedGraphs.ne(G)+N))
```

tsano430 commented on src/Metis.jl:
```diff
@@ -82,6 +85,40 @@ function graph(G::LightGraphs.AbstractSimpleGraph)
+    xadj[1] = 1
+    adjncy = Vector{idx_t}(undef, 2*SimpleWeightedGraphs.ne(G))
+    vwgt = ones(idx_t, N)
+    adjwgt = Vector{idx_t}(undef, 2*SimpleWeightedGraphs.ne(G))
```
Same as above.

Arjunayodya commented:
It would be helpful if some examples for performing node and edge-weighted grid partition are added. As Iam struggling with the creation of how SparseCSC can be created for creating node and edge weighted graphs. (like to sparse CSC for the required graph is created, adding edge and node weight, performing the metis)

Thanks in advance

fredrikekre commented:
#45 adds weight support for sparse matrix input. Shouldn't be too difficult to rebase this PR after that is merged.
